File: backend/app/schemas/stakeholder.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, EmailStr
from datetime import datetime

# Stakeholder Link type
class StakeholderLink(BaseModel):
    label: str
    url: str

# Stakeholder schemas
class StakeholderBase(BaseModel):
    """Base schema for Stakeholder"""
    name: str
    role: str
    organization: str
    type: str  # funder, implementer, mentor, policymaker, researcher, community_leader
    region: str
    email: EmailStr
    phone: Optional[str] = None
    bio: str
    tags: List[str] = []
    links: List[StakeholderLink] = []

class StakeholderCreate(StakeholderBase):
    """Schema for creating a new Stakeholder"""
    created_by: str

class StakeholderUpdate(BaseModel):
    """Schema for updating a Stakeholder"""
    name: Optional[str] = None
    role: Optional[str] = None
    organization: Optional[str] = None
    type: Optional[str] = None
    region: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    bio: Optional[str] = None
    tags: Optional[List[str]] = None
    links: Optional[List[StakeholderLink]] = None

class StakeholderResponse(BaseModel):
    """Schema for Stakeholder response"""
    id: int
    name: str
    role: str
    organization: str
    type: str
    region: str
    email: str
    phone: Optional[str] = None
    bio: str
    tags: List[str] = []
    links: List[StakeholderLink] = []
    created_by: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# StakeholderNote schemas
class StakeholderNoteBase(BaseModel):
    """Base schema for StakeholderNote"""
    content: str

class StakeholderNoteCreate(StakeholderNoteBase):
    """Schema for creating a new StakeholderNote"""
    created_by: str

class StakeholderNoteUpdate(BaseModel):
    """Schema for updating a StakeholderNote"""
    content: Optional[str] = None

class StakeholderNoteResponse(BaseModel):
    """Schema for StakeholderNote response"""
    id: int
    stakeholder_id: int
    content: str
    created_by: str
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

# Extended stakeholder response with notes
class StakeholderWithNotesResponse(StakeholderResponse):
    """Schema for Stakeholder response with notes"""
    notes: List[StakeholderNoteResponse] = []

File: frontend/src/types/stakeholder.ts
/**
 * Stakeholder Directory Types
 * Types for stakeholder records and associated notes
 */

export type StakeholderType =
  | 'funder'
  | 'implementer'
  | 'mentor'
  | 'policymaker'
  | 'researcher'
  | 'community_leader';

export interface StakeholderLink {
  label: string;
  url: string;
}

export interface Stakeholder {
  id: string;
  name: string;
  role: string;                    // Title
  organization: string;
  type: StakeholderType;
  region: string;                  // Country
  email: string;
  phone?: string;
  bio: string;
  tags: string[];
  links: StakeholderLink[];
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface Note {
  id: string;
  stakeholder_id: string;
  content: string;                 // Plain text / markdown
  created_by: string;
  created_at: string;
  updated_at: string;
}

// Form data types
export interface StakeholderFormData {
  name: string;
  role: string;
  organization: string;
  type: StakeholderType;
  region: string;
  email: string;
  phone?: string;
  bio: string;
  tags: string[];
  links: StakeholderLink[];
}

export interface NoteFormData {
  content: string;
}

// Filter types
export interface StakeholderFilters {
  search: string;
  type?: StakeholderType;
  region?: string;
}

// Display config for stakeholder types
export interface StakeholderTypeConfig {
  label: string;
  color: string;
  bgColor: string;
  icon: string;
}

export const stakeholderTypeConfigs: Record<StakeholderType, StakeholderTypeConfig> = {
  funder: {
    label: 'Funder',
    color: '#0077B6',
    bgColor: '#E8F1F9',
    icon: '💰'
  },
  implementer: {
    label: 'Implementer',
    color: '#A8E6CF',
    bgColor: '#F0FAF5',
    icon: '🎯'
  },
  mentor: {
    label: 'Mentor',
    color: '#FFD93D',
    bgColor: '#FFF9E5',
    icon: '🧑‍🏫'
  },
  policymaker: {
    label: 'Policymaker',
    color: '#C3B1E1',
    bgColor: '#F3EFFA',
    icon: '🏛️'
  },
  researcher: {
    label: 'Researcher',
    color: '#6BCF7F',
    bgColor: '#E8F8EB',
    icon: '🔬'
  },
  community_leader: {
    label: 'Community Leader',
    color: '#FF6B6B',
    bgColor: '#FFE8E8',
    icon: '👥'
  }
};
